### PR TITLE
Vortex build paul

### PR DESCRIPTION
**Pull request by william76**
@prwolfe 
Creating a PR here to track changes for when I get my analysis done on the vortex_build branch errors.

Unfortunately, I cannot log into vortex but I have a PR in.

In the meantime, if you get a chance, can you pop onto Vortex and check the modules and make sure that `StdEnv` is there?

Here's the error message I'm seeing currently with the added error messaging I added in this PR:
```
00:00:58.994 +--------------------------------------------------------------------+
00:00:58.994 |   E N V I R O N M E N T   S E T   U P   S T A R T
00:00:58.994 +--------------------------------------------------------------------+
00:00:58.994 Module Operations
00:00:58.994 +---------------+
00:00:58.994 [module] module load StdEnv **INTERNAL ERROR**
00:00:58.994 Debugging Information:
00:00:58.994 :  op     = load
00:00:58.994 :  params = ['StdEnv']
00:00:58.994 :     len(params) = 1
00:00:58.994 :  status = 0
00:00:58.996 Traceback (most recent call last):
00:00:58.996   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 259, in <module>
00:00:58.996     status = main(args)
00:00:58.996   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 249, in main
00:00:58.996     pr_config.prepare_test()
00:00:58.996   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/TrilinosPRConfigurationBase.py", line 563, in prepare_test
00:00:58.996     rval = tr_config.apply(throw_on_error=True)
00:00:58.996   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/setenvironment/SetEnvironment.py", line 192, in apply
00:00:58.996     raise(err)
00:00:58.996   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/setenvironment/SetEnvironment.py", line 176, in apply
00:00:58.996     cmd_status = max(ModuleHelper.module(op, params[0]), status)
00:00:58.996 TypeError: '>' not supported between instances of 'int' and 'NoneType'
00:00:59.015 Build step 'Execute shell' marked build as failure
00:00:59.318 Archiving artifacts
00:00:59.390 [Checks API] No suitable checks publisher found.

00:00:59.733 Finished: FAILURE

```

### CODE REVIEW

**Comment by william76** [comment]
oh, don't merge yet... this is more informative than 'done' for what I'm troubleshooting :p

**Comment by william76** [comment]
@prwolfe ok, I did some digging and it looks like `env_modules_python` comes from **lmod**.  It's a tiny file, here's the source:

```python
# -*- python -*-
from __future__ import print_function
from subprocess import PIPE, Popen
import os, sys


def module(command, *arguments):
    numArgs = len(arguments)
    A = ['/usr/local/Cellar/lmod/8.4.6/libexec/lmod', 'python', command]
    if (numArgs == 1):
        A += arguments[0].split()
    else:
        A += list(arguments)

    proc           = Popen(A, stdout=PIPE, stderr=PIPE)
    stdout, stderr = proc.communicate()
    err_out=sys.stderr
    if (os.environ.get('LMOD_REDIRECT','no') != 'no'):
        err_out=sys.stdout

    print(stderr.decode(),file=err_out)
    exec(stdout.decode())
```

it looks like it has no return value, which is why we're getting that `NoneType` error.  My suspicion is that we're using that one instead of our own in-house rolled version.

From the looks of that file, I'm not really sure how we'd be extracting the status of the command.  

I'm thinking one of two options is best here:
1. Remove our call to `env_modules_python` since it's generally not in many places and it appears to _not_ provide much I can use to check if our module commands are successful, etc.
2. Adjust `SetEnvironment` to assume the command was always successful if we're using the lmod `module` command since we can't really check success or failure.

I'm thinking (1) is the better option... what do you think?

**Comment by william76** [comment]
Confirmed.  I added a `Notice` line to the ModuleHelper.py file to tell us which `module()` is getting loaded, and I'm seeing this in the log file for [build 30][1]:
```
00:00:55.362 NOTICE> [ModuleHelper.py] Using the lmod based `env_modules_python` module handler.
```

I'll push an update that forces our version run another test -- I think our homegrown `module()` is probably better than the one LMOD has, so we might want to just remove that 'optional' dependency all together rather than try and support both.

[1]: https://ascic-jenkins.sandia.gov/job/trilinos-folder/job/Trilinos_pullrequest_cuda_10.1.243/30/console

**Comment by prwolfe** [comment]
Yeah – basically just assume if it didn’t raise an exception you are fine.

Paul

**Comment by prwolfe** [comment]
> In the meantime, if you get a chance, can you pop onto Vortex and check the modules and make sure that `StdEnv` is there?

Yes it is - in fact that is loaded by default I believe

**Comment by william76** [comment]
The latest run:
https://ascic-jenkins.sandia.gov/job/trilinos-folder/job/Trilinos_pullrequest_cuda_10.1.243/33/console

There may be a couple of cleanup items to take care of later once this is merged in regarding the ModuleHelper file... I think the right thing to do is to ditch the LMOD based module loader at this point in our routine since it feels like adding workarounds to that function's API is adding technical debt, especially since it's hard to track down info on that package. I had to go find the source of it to confirm it has no return value and look at side effects.

The build is getting past that traceback now but there are other errors that are popping up, but they appear to be more platform-specific and not so much an issue with the PR driver code:

<details>
<summary>Click to expand console log snippet</summary>
<pre>
00:02:26.669 +--------------------------------------------------------------------+
00:02:26.669 |   E N V I R O N M E N T   S E T   U P   S T A R T
00:02:26.669 +--------------------------------------------------------------------+
00:02:26.669 Module Operations
00:02:26.669 +---------------+
00:02:26.669 [module] module load StdEnv
00:02:26.669 [module output start]
00:02:26.669 
00:02:26.669 [module output end]
00:02:26.669 [module stderr start]
00:02:26.669 ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'StdEnv'
00:02:26.669 
00:02:26.669 [module stderr end]
00:02:26.669 
00:02:26.824  FAILED
00:02:26.824 [module] module load sparc-dev/cuda-10.1.243_gcc-7.3.1_spmpi-rolling
00:02:26.824 [module output start]
00:02:26.824 os.environ['ADIAK_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/adiak-bbbd526/31705b2a70193823115ce13a7a232dfde8639d25/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['BOOST_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/boost-1.72.0/d18f2d4458c316372ac45a6ea86cc53f3c9f2cb0/ats2-v100_cuda-10.1.243_gcc-7.3.1'
00:02:26.824 os.environ['CALIPER_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/caliper-b5315c1/2f5718e2b4bffea3c76d71e35fce53e350ff3bce/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['CBLAS_ROOT'] = '/usr/tcetmp/packages/lapack/lapack-3.8.0-gcc-4.9.3'
00:02:26.824 os.environ['CGNS_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/cgns-c09a5cd/1ca2cde5e84c243a87e46b5a90c2cca989d72959/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['CUDA_LAUNCH_BLOCKING'] = '1'
00:02:26.824 os.environ['CUDA_MANAGED_FORCE_DEVICE_ALLOC'] = '1'
00:02:26.824 os.environ['EUCLID_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/euclid-20.23/49b290856f56828303cd3e7a427721b4d2cc571c/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['HDF5_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/hdf5-1.10.5/ab761077d16a40b34e290223303deefd0c46b342/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['LAPACK_ROOT'] = '/usr/tcetmp/packages/lapack/lapack-3.8.0-gcc-4.9.3'
00:02:26.824 os.environ['LLNL_USE_OMPI_VARS'] = 'y'
00:02:26.824 os.environ['LOADEDMODULES'] = 'xl/2020.09.17:spectrum-mpi/rolling-release:cuda/10.1.243:StdEnv:git/2.20.0:python/3.7.2:sparc-tools/aerotools/2:sparc-tools/taos/2020.09.04:sparc-dev/cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['METIS_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/parmetis-4.0.3/bbc4da521d670fd9137b3c1e6917090ae52f91af/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['NETCDF_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/netcdf-4.7.0/5bcad755c39fe2155c36f6de0430790f852f4f33/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['OMPI_CXX'] = 'nvcc_wrapper'
00:02:26.824 os.environ['PARMETIS_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/parmetis-4.0.3/bbc4da521d670fd9137b3c1e6917090ae52f91af/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['PATH'] = '/usr/gapps/sparc/tools/nvcc_wrapper:/usr/gapps/sparc/tpls/ats2-v100/trilinos-develop/2020.08.01/b1667cfa76e01a632ff2d03754ade61813f8ebb0/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling_serial_static_opt/bin:/usr/gapps/sparc/tpls/ats2-v100/netcdf-4.7.0/5bcad755c39fe2155c36f6de0430790f852f4f33/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling/bin:/usr/gapps/sparc/tpls/ats2-v100/pnetcdf-1.12.1/f0cd40f050c0d85571bfb81d5c4f648874a10082/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling/bin:/usr/gapps/sparc/tpls/ats2-v100/cgns-c09a5cd/1ca2cde5e84c243a87e46b5a90c2cca989d72959/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling/bin:/usr/gapps/sparc/tpls/ats2-v100/hdf5-1.10.5/ab761077d16a40b34e290223303deefd0c46b342/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling/bin:/usr/gapps/sparc/tpls/ats2-v100/boost-1.72.0/d18f2d4458c316372ac45a6ea86cc53f3c9f2cb0/ats2-v100_cuda-10.1.243_gcc-7.3.1/bin:/projects/sparc/tools/ats2/clang-8.0.1/bin:/usr/gapps/sparc/tools/taos/2020-09-04/taos_ats2-pwr9_gcc-4.9.3_opt/bin:/usr/tce/packages/python/python-3.7.2/bin:/usr/tce/packages/git/git-2.20.0/bin:/usr/local/cuda/bin:/usr/tce/packages/cuda/cuda-10.1.243/bin:/usr/tce/packages/spectrum-mpi/spectrum-mpi-rolling-release-xl-2020.09.17/bin:/usr/tce/packages/xl/xl-2020.09.17/bin:/usr/tcetmp/bin:/usr/local/bin:/usr/bin:/opt/ibm/bb/bin:/opt/ibm/flightlog/bin:/opt/ibutils/bin:/usr/local/sbin:/usr/sbin:/usr/local/cuda-9.2/bin:/usr/lpp/mmfs/bin:/opt/ibm/spectrum_mpi/bin:/opt/ibm/spectrum_mpi/jsm_pmix/bin:/ascldap/users/trilinos/.local/bin:/ascldap/users/trilinos/bin:/opt/ibm/spectrumcomputing/lsf/10.1.0.9/linux3.10-glibc2.17-ppc64le-csm/etc:/opt/ibm/spectrumcomputing/lsf/10.1.0.9/linux3.10-glibc2.17-ppc64le-csm/bin:/usr/gapps/sparc/tools/aerotools/aeroxfer:/usr/gapps/sparc/tools/aerotools/csv2inp:/usr/gapps/sparc/tools/aerotools/pyromancer:/usr/gapps/sparc/tools/aerotools/redimensioner:/usr/gapps/sparc/tools/aerotools/scram:/usr/gapps/sparc/tools/aerotools/utils'
00:02:26.824 os.environ['PNETCDF_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/pnetcdf-1.12.1/f0cd40f050c0d85571bfb81d5c4f648874a10082/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['SGM_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/sgm-20.23/642d571ff38721f82cb08c2d186a9c283b92da38/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['SUPERLUDIST_ROOT'] = '/usr/gapps/sparc/tpls/ats2-v100/superlu_dist-5.4.0/29c9292a132e50b15c6723695c4a5b74b8eab987/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['TMPDIR'] = '/tmp/trilinos'
00:02:26.824 os.environ['TRILINOS_BASE'] = '/usr/gapps/sparc/tpls/ats2-v100/Trilinos/2020-07-28/00000000/ats2-v100_cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 os.environ['_LMFILES_'] = '/usr/tcetmp/modulefiles/Core/xl/2020.09.17.lua:/usr/tce/modulefiles/Compiler/xl/2020.09.17/spectrum-mpi/rolling-release.lua:/usr/tcetmp/modulefiles/Core/cuda/10.1.243.lua:/usr/tcetmp/modulefiles/Core/StdEnv.lua:/usr/tcetmp/modulefiles/Core/git/2.20.0.lua:/usr/tcetmp/modulefiles/Core/python/3.7.2.lua:/projects/sparc/modules/ats2/sparc-tools/aerotools/2:/projects/sparc/modules/ats2/sparc-tools/taos/2020.09.04:/projects/sparc/modules/ats2/sparc-dev/cuda-10.1.243_gcc-7.3.1_spmpi-rolling'
00:02:26.824 
00:02:26.824 [module output end]
00:02:26.824 [module stderr start]
00:02:26.824 python(7):ERROR:105: Unable to locate a modulefile for 'python/2.7.16'
00:02:26.824 cmake(17):ERROR:105: Unable to locate a modulefile for 'cmake/3.14.5'
00:02:26.824 xl/2020.09.17(1):ERROR:102: Tcl command execution failed: -- -*- lua -*-
00:02:26.824 
00:02:26.824 gcc(2):ERROR:105: Unable to locate a modulefile for 'gcc/7.3.1'
00:02:26.824 spectrum-mpi/rolling-release(1):ERROR:102: Tcl command execution failed: -- -*- lua -*-
00:02:26.824 
00:02:26.824 spectrum-mpi(2):ERROR:105: Unable to locate a modulefile for 'spectrum-mpi/rolling-release'
00:02:26.824 cuda/10.1.243(1):ERROR:102: Tcl command execution failed: -- -*- lua -*-
00:02:26.824 
00:02:26.824 cuda(2):ERROR:105: Unable to locate a modulefile for 'cuda/10.1.243'
00:02:26.824 lapack(11):ERROR:105: Unable to locate a modulefile for 'lapack/3.8.0-gcc-4.9.3'
00:02:26.824 
00:02:26.824 [module stderr end]
00:02:26.824 
00:02:26.851  FAILED
00:02:26.851 [module] module swap gcc/7.3.1 gcc/8.3.1Failed to execute the module command: swap gcc/7.3.1 gcc/8.3.1.
00:02:26.851 - Returned 1 exit status.
00:02:26.851 
00:02:26.851 [module output start]
00:02:26.851 
00:02:26.851 [module output end]
00:02:26.851 [module stderr start]
00:02:26.851 ModuleCmd_Switch.c(172):ERROR:152: Module 'gcc/7.3.1' is currently not loaded
00:02:26.851 
00:02:26.851 [module stderr end]
00:02:26.851 
00:02:26.852  FAILED
00:02:26.852 
00:02:26.852 Environment Vars
00:02:26.852 +--------------+
00:02:26.852 [envvar] export OMPI_CXX=${WORKSPACE}/Trilinos/packages/kokkos/bin/nvcc_wrapper
00:02:26.852 [envvar] export OMPI_CXX=/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/packages/kokkos/bin/nvcc_wrapper (actual)
00:02:26.852 [envvar] export OMPI_CC=${CC}
00:02:26.854 Traceback (most recent call last):
00:02:26.854   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 259, in <module>
00:02:26.854     status = main(args)
00:02:26.854   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 249, in main
00:02:26.854     pr_config.prepare_test()
00:02:26.854   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/TrilinosPRConfigurationBase.py", line 563, in prepare_test
00:02:26.854     rval = tr_config.apply(throw_on_error=True)
00:02:26.854   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/setenvironment/SetEnvironment.py", line 213, in apply
00:02:26.854     v = self._expand_envvars_in_string(v)
00:02:26.854   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/setenvironment/SetEnvironment.py", line 345, in _expand_envvars_in_string
00:02:26.854     raise KeyError(msg)
00:02:26.854 KeyError: 'Required environment variable `CC` does not exist.'
00:02:26.876 Build step 'Execute shell' marked build as failure
00:02:27.196 Archiving artifacts
00:02:27.259 [Checks API] No suitable checks publisher found.
00:02:27.529 Finished: FAILURE
</pre>
</details>

FYI @prwolfe

**Comment by prwolfe** [comment]
Those errors are mot right. We should not even be trying to load python/2.7.16 or cmake/3.14.5 (maybe they come from StdEnv?) I think the gcc error is fishy as well - our python module load may be non-functional.

**Comment by prwolfe** [comment]
I can confirm that we will need to use the Lua python module loader. Sad that is has no return value.

**Comment by prwolfe** [comment]
env_modules_python on vortex is even more simplistic that what you pasted above. The real issue seems to be that our version only looks for modulecmd and that is installed on vortex, but non-functional, while their version uses lmod instead. I see no reliable way to determine which command to use on any given platform (although mod is supposed to be backwards compatible while modulecmd does not understand Lua written modules at all.)

**Comment by william76** [comment]
> 
> 
> Those errors are mot right. We should not even be trying to load python/2.7.16 or cmake/3.14.5 (maybe they come from StdEnv?) I think the gcc error is fishy as well - our python module load may be non-functional.

Odd... we use that one on all our other platforms I think... otherwise we'd have seen the error with the NoneType comparison before.

The .ini looks more or less ok:
```
[Trilinos_pullrequest_cuda_10.1.243]
module-load StdEnv:
module-load sparc-dev: cuda-10.1.243_gcc-7.3.1_spmpi-rolling
module-swap gcc/7.3.1: gcc/8.3.1
setenv OMPI_CXX: ${WORKSPACE}/Trilinos/packages/kokkos/bin/nvcc_wrapper
setenv OMPI_CC: ${CC}
setenv OMPI_FC: ${FC}
setenv CUDA_LAUNCH_BLOCKING: 1
setenv CUDA_MANAGED_FORCE_DEVICE_ALLOC: 1
setenv PATH_TMP: /projects/atdm_devops/vortex/cmake-3.17.2/bin
setenv PATH:     /projects/atdm_devops/vortex/ninja-fortran-1.8.2:${PATH_TMP}
unsetenv PATH_TMP:
```
It looks like it couldn't load the `StdEnv` module and then the sparc-dev pack failed to load as well.

Hopefully I can get an account on that system tomorrow and go and have a look for myself. I will also instrument a little more for logging in ModuleHelper to figure out which `modulecmd` it's using since we're using the import module in a try..except block with default if not found trick. Here's the block of code where this is basically done:

https://github.com/prwolfe/Trilinos/blob/4395bf71352302a970718f391dc2da3361dd399f/cmake/std/trilinosprhelpers/setenvironment/ModuleHelper.py#L56-L75

we're really just executing a `modulecmd <args>` call via Popen...  I'm wondering if the issue is coming from line 69

```python
cmd = [modulecmd, "python"] + args
```

perhaps this should be changed to force "python3"?

**Comment by william76** [comment]
> 
> 
> env_modules_python on vortex is even more simplistic that what you pasted above. The real issue seems to be that our version only looks for modulecmd and that is installed on vortex, but non-functional, while their version uses lmod instead. I see no reliable way to determine which command to use on any given platform (although mod is supposed to be backwards compatible while modulecmd does not understand Lua written modules at all.)

Ah... ok, so on Vortex, they installed modulecmd but it's known to be broken so we really really want to use the lmod version?  

I can put something together to account for this but the lmod version of the `module()` function in Python gives me no way to determine if it was successful or failed. If you're at the command line on Vortex, do you just execute `module load <package>` or does it have a different command entirely?

**Comment by prwolfe** [comment]
> I can put something together to account for this but the lmod version of the `module()` function in Python gives me no way to determine if it was successful or failed. If you're at the command line on Vortex, do you just execute `module load <package>` or does it have a different command entirely?

the command line is still "module load xxx", but module is a shell function (on all machines) and on vortex that uses lmod where on the ascic machines and tlcc2 it uses mod. Fun isn't it?

In the long run, lmod is more stable than the tcl used by modulecmd, so I keep hoping more systems will move this way.

**Comment by william76** [comment]
@prwolfe 
I reverted that change to just disable using the lmod based `env_modules_python.module()` command, so we're back to the problem that we can't compare an int and a NoneType value.

I was able to get onto Vortex today and, here's the fun part, the `module()` function that is on that system in the `env_modules_python` is quite a bit different from what I'm seeing from the lmod that got installed on my laptop that I showed [above in the comments][1]. 
On Vortex, here's what is in `env_modules_python.py`
```python
# -*- python -*-
import os
def module(command, *arguments):
  commands = os.popen('/usr/share/lmod/lmod/libexec/lmod python %s %s'\
                      % (command, ' '.join(arguments))).read()
  exec(commands)
```
ehhhh... that's all of it.  /facepalm.  so they call popen with absolutely no checking or anything to determine success or failure of the command. I find it kind of amazing that they return **no state** for something like a module load command since that's exactly the kind of command you might want to verify that it worked inside Python. **Note: you can find that `env_modules_python.py` via `$ ls ${MODULESHOME}/init`**

So here's a bit of a problem... the _modules_ based `module()` function we define if we can't find the lmod based `module()` command is kinda smart and it returns a value from the function call: 0 if successful and > 0 otherwise. The code in `SetEnvironment` relies on this... so the problem here is the lack of a return value coming from the function in `SetEnvironment` which is where we're getting the traceback about he comparison between NoneType and int.

Are you around tomorrow (Friday)?  I have a couple ideas on how to attack this that we can discuss... generally speaking the options I'm weighing are:
1. Add a `module()` wrapper to the lmod based command inside `ModuleHelper` that always just returns an int 0 and trundle along. This might be the _easiest_ thing to do but then we lose the ability to determine if the module operations were successful or not.
2. Modify the checks in `SetEnvironment` to catch the `NoneType` return value from the calls to `module()` and only do the success check if we got an integer back (i.e., only check success if we used our version of the modules command). I kind of hate this because it feels like the kind of thing that adds technical debt as a workaround for a deficiency at a lower level.
3. Implement (1) but also extend the 'language' that `SetEnvironment` knows from its .ini file to add in some ENVVAR validation commands. The idea here is that if we can't verify that the module command exited with a particular status then can we add verification steps to our configuration files that would fail our build if we are missing envvars that we expect the module to load if it was successful? Possibly a couple variations to validate the existence of an envvar only up through matching the value of an envvar via string match or regex... 

Right now those are the main options I'm weighing. I like (1) or (3) better than (2).  Option 3 might take a bit of work depending on how far we want to get into it but it might be the most robust since apparently there is a significant variation in what lmod puts into their `module()` function.

Doing (1) is pretty trivial to perform and will be the least impactful but without (3) we won't have any way to verify that the modules loaded... but getting that in quickly might help you make more progress on the Vortex configuration.

Your thoughts?  Shoot me a meeting invite tomorrow if you're around and want to kibbitz on it.

FYI @jwillenbring 

[1]: https://github.com/prwolfe/Trilinos/pull/15#issuecomment-725591145

**Comment by prwolfe** [comment]
Option 1 sounds fine to me. You might wrap it in a try/except as failure should raise an exception. That's the only failure indicator you are going to get.

**Comment by william76** [comment]
@prwolfe 
I may have it fixed up at this point as far as the `module()` command goes:

https://ascic-jenkins.sandia.gov/job/trilinos-folder/job/Trilinos_pullrequest_cuda_10.1.243/38/console
```
00:00:53.314 packageEnables Command: 
00:00:53.314 $ /vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/commonTools/framework/get-changed-trilinos-packages.sh \
00:00:53.314     origin/vortex_build_paul \
00:00:53.314     HEAD \
00:00:53.314     packageEnables.cmake \
00:00:53.314     package_subproject_list.cmake
00:00:53.314 
00:00:53.327 /vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/commonTools/framework/get-changed-trilinos-packages.sh: line 81: readlink: command not found
00:00:53.327 Could not follow symlink to set TRILINOS_DIR!
00:00:53.328 ERROR: Cannot determine TRILINOS_DIR!  Please set env var TRILINOS_DIR!
00:00:53.328 --- There was an issue generating `packageEnables.cmake`.
00:00:53.328 --- The error code was: 4
00:00:53.328 --- Console Output:
00:00:53.328 None
00:00:53.335 Traceback (most recent call last):
00:00:53.335   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 259, in <module>
00:00:53.335     status = main(args)
00:00:53.335   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 249, in main
00:00:53.335     pr_config.prepare_test()
00:00:53.335   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/TrilinosPRConfigurationBase.py", line 604, in prepare_test
00:00:53.335     self.create_package_enables_file(dryrun=self.args.dry_run)
00:00:53.335   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/TrilinosPRConfigurationBase.py", line 439, in create_package_enables_file
00:00:53.335     raise cpe
00:00:53.335   File "/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/cmake/std/trilinosprhelpers/TrilinosPRConfigurationBase.py", line 431, in create_package_enables_file
00:00:53.335     subprocess.check_call(cmd)
00:00:53.335   File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/subprocess.py", line 347, in check_call
00:00:53.335     raise CalledProcessError(retcode, cmd)
00:00:53.335 subprocess.CalledProcessError: Command '['/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_10.1.243/Trilinos/commonTools/framework/get-changed-trilinos-packages.sh', 'origin/vortex_build_paul', 'HEAD', 'packageEnables.cmake', 'package_subproject_list.cmake']' returned non-zero exit status 4.
00:00:53.357 Build step 'Execute shell' marked build as failure
00:00:53.875 Archiving artifacts
00:00:53.946 [Checks API] No suitable checks publisher found.
00:00:54.379 Finished: FAILURE
```

I added a simple wrapper for the lmod version:

https://github.com/prwolfe/Trilinos/blob/1be7540fb35f16da67b262d2a7c1b07b13c7f3a5/cmake/std/trilinosprhelpers/setenvironment/ModuleHelper.py#L29-L63

It just returns 0, but if an exception happens I'll pass that along.

and also updated the default ModuleHelper 'modules' based handler to use the same kind of interface as the LMOD one is for consistency.

https://github.com/prwolfe/Trilinos/blob/1be7540fb35f16da67b262d2a7c1b07b13c7f3a5/cmake/std/trilinosprhelpers/setenvironment/ModuleHelper.py#L71

I think I may want to rework the ModuleHelper.py file sometime to make something a bit more modular.

This can probably be merged into your fork now.

**Comment by william76** [comment]
@prwolfe 
Update: I added a setenv to create the envvar TRILINOS_DIR to the .ini file for the cuda 10.1.243 job.
```
setenv TRILINOS_DIR: ${WORKSPACE}/Trilinos
```
https://ascic-jenkins.sandia.gov/job/trilinos-folder/job/Trilinos_pullrequest_cuda_10.1.243/39/console

It looks like we're moving along... the latest error I'm seeing looks like we don't have a `git` module loaded.

I'll go ahead and hand the rest of this off to you... Let me know when we get this merged into Develop so I can pull in the updates I've made to the setenvironment module :D

**Review by prwolfe — APPROVED** [review]

**Pull request merged**
